File: fetchETHprice.py
from flask import Flask, jsonify
from flask_cors import CORS
import requests

app = Flask(__name__)
CORS(app)

# Define the Moralis API URL
MORALIS_API_URL = "https://deep-index.moralis.io/api/v2/erc20/price?chain=eth&address=0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee"
MORALIS_API_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJub25jZSI6IjJ5ZDMwLWJWYWJwZ1YyYTZndTNrMzg1M05lMyIsInVvY2tldCI6ImIzNzUxN2U3LWVkY2UtNDI1Mi04NTY5LWJiMjcwNzU3NTVlMyIsIm9yZ0lkIjoiMzY1Iiwib3JnUm9sZSI6IlNlcnZlciIsImlhdCI6MTcxMTQ1NDY2MywiZXhwIjoxNzEzNjEwNjYzfQ.oRz0-4R9JvQ6M5Vz3mLwQyzjAYvzhIMDJkMSIS"
USD_TO_KSH = 129  # Fixed exchange rate


@app.route('/get_eth_price')
def get_eth_price():
    try:
        print("Fetching ETH price from Moralis...")  # Add this line

        # Fetch ETH price in USD from Moralis
        headers = {"X-API-Key": MORALIS_API_KEY}  # Corrected headers
        print(f"Request Headers: {headers}")
        response = requests.get(MORALIS_API_URL, headers=headers)

        print(f"Response Status Code: {response.status_code}") # Add this line
        print(f"Response Text: {response.text}") # Add this line

        if response.status_code == 200:
            data = response.json()
            print(f"Data from Moralis: {data}")  # Add this line
            eth_price_usd = data.get("usdPrice", 0)
            print(f"ETH Price in USD: {eth_price_usd}")  # Add this line

            # Convert to KSh
            eth_price_ksh = eth_price_usd * USD_TO_KSH

            return jsonify({"eth_price_ksh": eth_price_ksh})
        else:
            print("Failed to fetch Ethereum price from Moralis")
            return jsonify({"error": "Failed to fetch Ethereum price"}, 500)

    except Exception as e:
        print(f"An exception occurred: {str(e)}") #Add this line
        return jsonify({"error": str(e)}, 500)


if __name__ == '__main__':
    app.run(debug=True, port=5000)

File: fetchETHPrice.js
async function fetchETHPrice() {
    try {
        // Get the amount entered by the user
        const ethAmount = parseFloat(document.getElementById('cryptoAmount').value) || 1;
        
        // Fetch ETH price from the backend
        const response = await fetch('http://127.0.0.1:5000/eth-price');
        const data = await response.json();
        
        if (data.eth_price_ksh) {
            // Display the price of 1 ETH
            document.getElementById('price-display').textContent =
                `1 Ethereum is approximately ${data.eth_price_ksh.toLocaleString()} KSh`;
            
            // Calculate and display the converted amount
            const convertedAmount = ethAmount * data.eth_price_ksh;
            document.getElementById('result').textContent =
                `${ethAmount} Ethereum is approximately ${convertedAmount.toLocaleString()} KSh`;
        } else {
            document.getElementById('price-display').textContent = 'Error fetching ETH price';
            document.getElementById('result').textContent = 'Failed to convert ETH to KSh';
        }
    } catch (error) {
        console.error('Error fetching ETH price:', error);
        document.getElementById('price-display').textContent = 'Failed to fetch Ethereum price.';
        document.getElementById('result').textContent = 'Error: ' + error.message;
    }
}
